### PR TITLE
E2222. Refactor impersonate_controller.rb

### PR DESCRIPTION
**Pull request by moksh98**
Please set the title of your pull request in the following format: `[Project id/Independent study/Issue fix/...]. project name`, eg., `E904. Integrating courses and assignments on Expertiza with Moodle`

We suggest you including the below information in your pull request:
- A description of the changes proposed in the pull request.
- @mentions of the person or team responsible for reviewing proposed changes.
---
About Expertiza Bot
- If you have any questions about comments given by the expertiza-bot [(example)](https://github.com/expertiza/expertiza/pull/1877#issuecomment-762412918), you could create a comment in your pull request with the format `/dispute [UUID1] [UUID2]` [(example)](https://github.com/expertiza/expertiza/pull/1877#issuecomment-762430057) then the professor and TAs will be notified.
- [Here is a video](https://tinyurl.com/internet-bots) about how to communicate to expertiza-bot.

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -84,6 +84,8 @@ GEM
       activesupport (>= 4.2, < 5.3)
```
Comment (namanshrimali) [inline Gemfile.lock:84]: Get rid of the changes in Gemfile.lock.

File: app/controllers/impersonate_controller.rb [hunk 2]
```diff
@@ -104,40 +112,34 @@ def display_error_msg
...
-    end
+  def do_impersonate_operation(user)
+    check_if_user_impersonateable if user
+    display_error_msg
   end
 
   # Main operation, method used to break the functions in impersonate controller and bring out 2 functionalities at same level,
   # checking if user impersonateable, if not throw corresponding error message
+
   def impersonate
```
Comment (namanshrimali) [inline app/controllers/impersonate_controller.rb:127]: This entire impersonate method is poorly written. `display_error_msg` method is terrible, and I believe should be removed entirely. This is how I think this logic should be implemented. You can check the implementation present in the master branch and take inspiration from there as that is working as expected.
* Implement another method that checks the input params and should show error if parameters are invalid (for example, existence of special characters or if field is empty. This method should be called before the impersonate is called. You can check out how `before_action` can be used for this case.
* After the input is checked, we can go on with our logic

File: app/controllers/impersonate_controller.rb [hunk 3]
```diff
@@ -79,18 +84,21 @@ def check_if_user_impersonateable
...
-        # E1991 : check whether instructor is currently in anonymized view
         overwrite_session
       end
     end
   end
 
   # Function to display appropriate error messages based on different username provided, the message explains each error
+  # This function checks params values and displays error_message based on the user name .This is initial check to see if username is valid
+
   def display_error_msg
```
Comment (namanshrimali) [inline app/controllers/impersonate_controller.rb:95]: We'll be working to eliminate this method.

File: app/controllers/impersonate_controller.rb [hunk 4]
```diff
@@ -1,6 +1,9 @@
 class ImpersonateController < ApplicationController
   include SecurityHelper
 
+  # This function checks if the logged in user is a student or not. If it is a student, do not allow the impersonate mode.
+  # If the logged in user has the role or anything other than the student, we allow that user to use the impersonate mode.
+
   def action_allowed?
```
Comment (namanshrimali) [inline app/controllers/impersonate_controller.rb:7]: You can look over the autocomplete guide from [here](https://www.jquerycards.com/forms/autocomplete/rails4-autocomplete/)

File: Gemfile.lock (hunk 1 above)
Comment (moksh98) [reply Gemfile.lock]: Resolved

File: app/controllers/impersonate_controller.rb [hunk 5]
```diff
@@ -104,40 +112,34 @@ def display_error_msg
     end
   rescue StandardError
     flash[:error] = @message
-    redirect_to :back
+    redirect_back
```
[the remark this answers is not in the document]
Comment (moksh98) [reply app/controllers/impersonate_controller.rb]: Resolved

File: app/controllers/impersonate_controller.rb [hunk 6]
```diff
@@ -159,4 +161,15 @@ def impersonate
       redirect_to :back
```
[the remark this answers is not in the document]
Comment (moksh98) [reply app/controllers/impersonate_controller.rb]: resolved

File: app/controllers/impersonate_controller.rb [hunk 7]
```diff
@@ -13,63 +16,65 @@ def action_allowed?
...
+  # can_impersonate method in user.rb checks whether the original user can impersonate the other user in params
+  # This method checks whether the user is a superadmin or teaching staff or recursively adds the child users till it reached highest hierarchy which is SuperAdmin
+  # If original user can impersonate the user ,then session will be overwrite to get the view of the user who is getting impersonated
+
   def check_if_user_impersonateable
     if params[:impersonate].nil?
-      # E1991 : check whether instructor is currently in anonymized view
-      user = User.anonymized_view?(session[:ip]) ? User.real_user_from_anonymized_name(params[:user][:name]) : User.find_by(name: params[:user][:name])
+      user = get_real_user(params[:user][:name])
       if !@original_user.can_impersonate? user
```
[the remark this answers is not in the document]
Comment (moksh98) [reply app/controllers/impersonate_controller.rb]: resolved

File: app/controllers/impersonate_controller.rb (hunk 3 above)
Comment (moksh98) [reply app/controllers/impersonate_controller.rb:93]: should we do anything to this method?

Comment (namanshrimali) [reply app/controllers/impersonate_controller.rb:95]: By "we" I meant your team and I xD

File: app/controllers/impersonate_controller.rb (hunk 2 above)
Comment (vijdurk) [reply app/controllers/impersonate_controller.rb:134]: Resolved.
Created a method to check the validity of params . Using before_action ,this method is first called before the impersonate method. Removed the display_error_message method and its dependencies in the code

File: app/controllers/impersonate_controller.rb (hunk 3 above)
Comment (vijdurk) [reply app/controllers/impersonate_controller.rb]: Removed this method and its dependencies in the code

**Pull request merged**